core/contextmenu_items.js

```diff
@@ -270,12 +270,230 @@ Blockly.ContextMenuItems.registerWorkspaceOptions_ = function() {
   Blockly.ContextMenuItems.registerDeleteAll();
 };
 
+/** Option to duplicate a block. */
+Blockly.ContextMenuItems.registerDuplicate = function() {
+  var duplicateOption = {};
+  duplicateOption.displayText = function() {
+    return Blockly.Msg['DUPLICATE_BLOCK'];
+  };
+  duplicateOption.preconditionFn = function(scope) {
```

Comment:
For here and below, set scope type: 
```
function (/** @type {!Blockly.ContextMenuRegistry.Scope} */ scope) {
...
}
```

Reply:
This will help catch things like use of protected / private properties.

Reply:
Done!